### PR TITLE
Fix Prephase next entity and done state freeze

### PR DESCRIPTION
**Pull request by pakfront**
DRAFT: Fixes https://github.com/MegaMek/megamek/issues/4291 where prephase (Premovement and Prefiring)
Next correctly searches unit list for next not-done and not-revealed unit. Switched order of entity update and current turn update so entity sent first.

### CODE REVIEW

**Review by SJuliez — COMMENTED** [review]
Comments...

File: megamek/src/megamek/client/ui/swing/PrephaseDisplay.java [hunk 1]
```diff
@@ -345,22 +338,12 @@ private void beginMyTurn() {
 
         clientgui.getBoardView().clearFieldofF();
 
+        selectEntity(clientgui.getClient().getFirstHiddenEntityNum());
```
[the remark this answers is not in the document]
Comment (pakfront) [reply megamek/src/megamek/client/ui/swing/PrephaseDisplay.java:341]: Currently it is safe, as you should get into this area if you have no hidden units. I tested with one player hacing hiden units and the other not having.

File: megamek/src/megamek/common/Game.java [hunk 2]
```diff
@@ -1816,6 +1816,42 @@ public int getNextDeployableEntityNum(GameTurn turn, int start) {
         return getFirstDeployableEntityNum(turn);
     }
 
+    /**
+     * @param turn the current game turn, which may be null
+     * @return the number of the first hidden entity that is valid for the specified turn
+     */
+    public int getFirstHiddenEntityNum(final @Nullable GameTurn turn) {
+        // Reviewers: Not sure if this is where to add filtering (this is hoe deployment does it)
```
[the remark this answers is not in the document]
Comment (pakfront) [reply megamek/src/megamek/common/Game.java:1824]: I'll try it out with a new GameTurn subclass. It "feels" like the right way to do it. If it works, we should add a ticket to make deployment work in the same way if possible.

[the remark this answers is not in the document]
Comment (pakfront) [reply megamek/src/megamek/common/Game.java:1824]: @SJuliez https://github.com/MegaMek/megamek/pull/4297/commits/1a6ddc43191e75b99df3324312c84f3ab872d63d uses the new PrephaseTurn subclass of GameTurn. Works well.
I could not add it to a new file, as the pattern for the other subclasses is a a static inner class of GameTurn. Not sure why that pattern was chosen, but it seemed ill-advised to deviate.

File: megamek/src/megamek/server/GameManager.java [hunk 3]
```diff
@@ -12751,13 +12751,17 @@
             return;
         }
 
+
+        entity.setDone(true);
```
[the remark this answers is not in the document]
Comment (rjhancock) [reply megamek/src/megamek/server/GameManager.java:12755]: Look into this to prevent possible null value

**Review by NickAragua — APPROVED** [review]
Any time someone messes around with the game turn stuff, I cross my fingers.

**Pull request merged**